Acceleration based control/feed-forward

MaEtUgR opened the pull request:
**Describe problem solved by this pull request**
More or less last step to finish splitting up and QAing PX4/Firmware#12072. It's based on #245 and introduces processing acceleration setpoints. This enables acceleration feed-forward which the jerk optimized flight tasks by @bresch immediately make use of. It also heavily reduces dependence between collective thrust and vehicle tilt making the demanded tilt less jerky when the altitude control output is changing.

**Describe your solution**
I make the velocity control output acceleration in m/s² which in theory would make the controller tuning more independent of the vehicle's thrust to weight ratio (~hover thrust) but for the initial introduction I introduce a scaling factor using the hover thrust to make the existing gains compatible. To separate the high-frequency altitude control changes from the collective thrust from the demanded tilt I'm switching control strategy to apply horizontal acceleration setpoints assuming hover/gravitational acceleration in the vertical direction effectively directly mapping it to the vehicle's tilt. The vertical acceleration gets projected onto the planned tilt and then executed by setting the collective thrust accordingly. Only in the limit case of more than the maximum thrust tilt gets coupled by being limited using vertical acceleration demand to ensure prioritizing altitude control over horizontal navigation.

**Test data / coverage**
For the basic limit cases, input combinations, failsafe and zero thrust inputs there are unit tests now. I did a lot of SITL testing and multiple outdoor tests with a 250 size, one test on a big multicopter.

**Additional context**
Refactorings that made this change possible: https://github.com/PX4/Firmware/pull/13247, https://github.com/PX4/Firmware/pull/13248, https://github.com/PX4/Firmware/pull/13262, https://github.com/PX4/Firmware/pull/13347, https://github.com/PX4/Firmware/pull/13701, https://github.com/PX4/Firmware/pull/14017

- [x] I still need to change the jerk limits for manual and auto since the tracking is tighter now.
- I discussed with @Jaeyoung-Lim to expose the acceleration on the offboard interface to allow more dynamic flights and a different kind of input that doesn't rely on vehicle internal velocity estimate.

MaEtUgR commented:
I rebased on master and removed the draft since the configuration default changes are also there.

Jaeyoung-Lim commented:
@MaEtUgR I tried offboard acceleration feedforward with this PR, but couldn't enter offboard mode. Am I missing something?

MaEtUgR commented:
@Jaeyoung-Lim Thanks for testing offboard!!
The provided logs show that instead of actually setting acceleration the normalized collective thrust vector is set to the value you probably want to give in m/s^2
![image](https://user-images.githubusercontent.com/4668506/75337009-7633e680-588c-11ea-9cd8-fa7a23c1359c.png)
I presumably fixed that with this pr: https://github.com/PX4/Firmware/pull/14212/files#diff-f58f83b125f08a9604e2b91bb9f94b1bL231-L233

I don't know why you can't enter offboard mode, maybe we can have a quick look together to get a clue?

Jaeyoung-Lim commented:
@MaEtUgR Ah, sorry for not being clear. That was the logs from master since I couldn't make it go in offboard mode. Would be great if we can dig in!

Jaeyoung-Lim commented:
@MaEtUgR Disarming shortly after arming in offboard mode : [Log](https://review.px4.io/plot_app?log=a398dabf-5fd8-4e35-9ebe-22cd68550229)

dannyfpv commented:
tested on pixhawk4 v5 f-450
Modes Tested*

Position Mode: Good.
Altitude Mode: Good.
Stabilized Mode: Good.
Mission Plan Mode (Automated): Good.
go to: Good.
RTL: Good.

- Procedure
Arm and Take off in position mode, after flying for approximately one minute, switched to altitude then stabilized mode proceed to switch to mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once completed all waypoin activate RTL

Notes:
No issues noted, good flight in general.

Log:
https://review.px4.io/plot_app?log=d2109f81-1651-4b5d-91de-92147ee43378

Junkim3DR commented:
### Tested on NXP FMUK66 v3

**Modes Tested**
- Mission Plan Mode (Automated): 
- Good.Go To Mode (Automated): Good.
- RTL (Return To Land): Good.

**Procedure**
**Note: Vehicle was sent to a _"Go To"_ point then RTL was triggered**
Arm and Take off in mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once all waypoints are completed make sure vehicle triggers RTL.

**Notes**
Good flight in general.

**Logs**
https://review.px4.io/plot_app?log=b3dfd093-4eb9-457b-9723-ddbe76de9ccf





### Tested on NXP FMUK66 v3

**Modes Tested**
- Mission Plan Mode (Automated): 
- Good.Go To Mode (Automated): Good.
- RTL (Return To Land): Good.

**Procedure**
**Note: Vehicle was sent to a _"Go To"_ point then RTL was triggered**
Arm and Take off in mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once all waypoints are completed make sure vehicle triggers RTL.

**Notes**
Good flight in general.

**Logs**
https://review.px4.io/plot_app?log=c8254650-1392-4e7a-94fc-de6d030a939f

jorge789 commented:
### Tested on PixRacer V4

**Tested Modes**

Position Mode
Mission plan mode (automated):
Good. Go to (Automated) mode: Good.
RTL (Back to Earth): Good.
**Process**
Note: the vehicle was sent to a "Go to" point in the middle of mission and then RTL was activated
Assemble and take off in Position mode after flying one minute, switch to mission plan, then make sure the vehicle follows all the waypoints as shown in QGC, once all waypoints are completed, make sure of the vehicle activating RTL.

**Notes**
Good flight overall.

**Logs**
https://review.px4.io/plot_app?log=7286d189-3cdf-4a76-aa01-2edfe1a6633d

https://review.px4.io/plot_app?log=6aa51b7b-f999-4aea-bcdd-1a4cb0bde0ac


### Tested on CUAV nano V5

**Tested Modes**

Position Mode
Mission plan mode (automated):
Good. Go to (Automated) mode: Good.
RTL (Back to Earth): Good.
**Process**
Note: the vehicle was sent to a "Go to" point in the middle of mission and then RTL was activated
Assemble and take off in Position mode after flying one minute, switch to mission plan, then make sure the vehicle follows all the waypoints as shown in QGC, once all waypoints are completed, make sure of the vehicle activating RTL.

**Notes**
Good flight overall.

**Logs**
https://review.px4.io/plot_app?log=ac76803a-1689-420b-9516-fb67c423614c

https://review.px4.io/plot_app?log=d12827ac-cfd1-48d1-babe-b85b10c54264

MaEtUgR commented:
@dagar As we discussed I rebased (without conflict) and removed the in my eyes deprecated `FlightTaskSport` to make it fit in the flash of the omnibus. Here's the commit message explaining:
```
This mode was just kept as an example after
its usage in a single case. It's basically untested
and doesn't make much sense anymore since it's
incompatible with the jerk limited trajectory
implementations. It's implementation only switched
hte configuration parameter of the velocity resulting
from maximum stick deflection to be
MPC_XY_VEL_MAX instead of MPC_VEL_MANUAL.
This is according to todays understanding undesired
because when hitting that limit the position
controller has no room for corrections anymore.

Also it saves some flash space on omnibus to remove
the task at this point and makes romm for the
acceleration feed-forward.
```

MaEtUgR commented:
I rebased on master and tried to reproduce the offboard issues by using the MAVSDK offboard integration tests:
`./build/src/integration_tests/integration_tests_runner --gtest_filter="SitlTest.Offboard*"`
But they all pass and seem to be doing what is expected looking at the simulation UI.

What do the two different errors I had come from, could they be glitches?
before rebasing: http://ci.px4.io:8080/blue/organizations/jenkins/PX4_misc%2FFirmware-SITL_tests/detail/PR-14212/6/pipeline
after rebasing: http://ci.px4.io:8080/blue/organizations/jenkins/PX4_misc%2FFirmware-SITL_tests/detail/PR-14212/7/pipeline

Now also jenkins seems to cancel my build NuttX builds only one shows in the UI and whole test fails even though the GitHub workflow builds succeed and "dronecode-macmini" cannot be contacted for the Mac build.

hamishwillee commented:
@MaEtUgR Any docs impact to all this?

MaEtUgR commented:
@hamishwillee Good point. Sport mode was never documented since it wasn't too different. But the control diagram https://dev.px4.io/master/en/flight_stack/controller_diagrams.html#multicopter-position-controller needs to be adjusted a little bit. Time to look into how these diagrams 😬

MaEtUgR commented:
I was able to reproduce what CI is flying and it found a real issue that I'm currently looking into.

MaEtUgR commented:
It is a problem with the scaling and anti-windup of the integrator, I need to find good unit tests to capture this and find a solution.

MaEtUgR commented:
I rebased on master and resolved conflicts with https://github.com/PX4/Firmware/commit/3049a3d14d0f4f0309cf324b9a6c3b64d4e1825c because I needed this change already for #312.

MaEtUgR commented:
I think I found the problem: The Rundqwist anti-windup requires to know the setpoint before and after limiting and the tilt limit is already applied in the "before setpoint" and hence gets neglected. I need to bring them under the same denominator. I might just revert the new limiting mechanism to get the execution working and then have new limiting in a separate pr, let me see

MaEtUgR commented:
I rebased on master and looked through it again and saw that the last corrections changed the shape of the integrator plots in an unexpected way. I removed them to have the following basically identical shape to master with just the horizontal integrator fix compared to when it was tested last:
![image](https://user-images.githubusercontent.com/4668506/77233077-64044a00-6ba5-11ea-8388-c02afd10e2b6.png)
I can add additional improvements in the next pr.

MaEtUgR commented on src/lib/flight_tasks/tasks/Transition/FlightTaskTransition.cpp:
```diff
@@ -64,7 +64,7 @@ void FlightTaskTransition::updateAccelerationEstimate()
 	// Estimate the acceleration by filtering the raw derivative of the velocity estimate
 	// This is done to provide a good estimate of the current acceleration to the next flight task after back-transition
-	_acceleration_setpoint = 0.9f * _acceleration_setpoint + 0.1f * (_velocity - _velocity_prev) / _deltatime;
+	_acceleration_setpoint = .9f * _acceleration_setpoint + .1f * (_velocity - _velocity_prev) / _deltatime;
```
@bresch I just saw the problem that this overwrites acceleration setpoints used to level out the vehicle during transition. These setpoints were not executed before and now they actually are. Can I remove this override of the setpoint?

MaEtUgR replied:
I added it for testing: https://github.com/PX4/Firmware/pull/14212/commits/4aaad24d2cdc05117c39720459c9ece4d16eaab2

bresch commented:
@MaEtUgR Since the integrator is in the velocity controller, could you add a plot that shows the velocity response, please? (I'm just curious to see how the step response looks like)

MaEtUgR commented:
Rebased on master and lowered the maximum jerk in the VTOL defaults based on @fury1895's valuable VTOL testing feedback 🎉 . That concludes testing and fixes from my side and I consider it ready to merge 👍

Jaeyoung-Lim commented:
@MaEtUgR Maybe this is too late, but I have tried this with offboard control, but this doesn't seem to work with offboard control.

MaEtUgR commented:
@Jaeyoung-Lim It's never too late. Is there an easy way to reproduce such that we can find the issue?

Jaeyoung-Lim commented:
@MaEtUgR You can use the ROS node I created, but if this is too much overhead I can also create a simple mavsdk app for testing

MaEtUgR commented:
> I can also create a simple mavsdk app for testing

I think that would be useful not just for me but also to put in CI to make sure it does not break again.
